src/test/java/io/confluent/connect/jdbc/sink/JdbcSinkTaskTest.java

```diff
@@ -291,4 +297,180 @@ void initWriter() {
+    expect(ctx.errantRecordReporter()).andReturn(reporter);
+    expect(reporter.report(anyObject(), anyObject())).andReturn(CompletableFuture.completedFuture(null));
+    mockWriter.closeQuietly();
+    expectLastCall();
+    replayAll();
+
+    Map<String, String> props = new HashMap<>();
+    props.put(JdbcSinkConfig.CONNECTION_URL, "stub");
+    props.put(JdbcSinkConfig.MAX_RETRIES, String.valueOf(maxRetries));
+    props.put(JdbcSinkConfig.RETRY_BACKOFF_MS, String.valueOf(retryBackoffMs));
+    task.start(props);
+
+    task.put(records);
+
+    verifyAll();
+  }
+
+  @Test
+  public void batchErrorReporting() throws SQLException {
+    final int maxRetries = 0;
+    final int retryBackoffMs = 1000;
+    final int batchSize = 3;
+
+    List<SinkRecord> records = new ArrayList<>();
+    SinkRecord record = new SinkRecord("stub", 0, null, null, null, null, 0);
+
+    for (int i = 0; i < batchSize; i++) {
+      records.add(record);
+    }
+
+    final JdbcDbWriter mockWriter = createMock(JdbcDbWriter.class);
+    SinkTaskContext ctx = createMock(SinkTaskContext.class);
+
+    mockWriter.write(records);
+    SQLException exception = new SQLException("cause 1");
+    expectLastCall().andThrow(exception);
+    mockWriter.write(anyObject());
+    expectLastCall().andThrow(exception).times(batchSize);
+
+    JdbcSinkTask task = new JdbcSinkTask() {
+      @Override
+      void initWriter() {
+        this.writer = mockWriter;
+      }
+    };
+    task.initialize(ctx);
+    ErrantRecordReporter reporter = createMock(ErrantRecordReporter.class);
+    expect(ctx.errantRecordReporter()).andReturn(reporter);
+    expect(reporter.report(anyObject(), anyObject())).andReturn(CompletableFuture.completedFuture(null)).times(batchSize);
+    for (int i = 0; i < 3; i++) {
+      mockWriter.closeQuietly();
+      expectLastCall();
+    }
+    replayAll();
+
+    Map<String, String> props = new HashMap<>();
+    props.put(JdbcSinkConfig.CONNECTION_URL, "stub");
+    props.put(JdbcSinkConfig.MAX_RETRIES, String.valueOf(maxRetries));
+    props.put(JdbcSinkConfig.RETRY_BACKOFF_MS, String.valueOf(retryBackoffMs));
+    task.start(props);
+
+    task.put(records);
+
+    verifyAll();
+  }
+
+  @Test
+  public void oneInBatchErrorReporting() throws SQLException {
+    final int maxRetries = 0;
+    final int retryBackoffMs = 1000;
+    final int batchSize = 3;
+
+    List<SinkRecord> records = new ArrayList<>();
+    SinkRecord record = new SinkRecord("stub", 0, null, null, null, null, 0);
+
+    for (int i = 0; i < batchSize; i++) {
+      records.add(record);
+    }
+
+    final JdbcDbWriter mockWriter = createMock(JdbcDbWriter.class);
+    SinkTaskContext ctx = createMock(SinkTaskContext.class);
+
+    mockWriter.write(records);
+    SQLException exception = new SQLException("cause 1");
+    expectLastCall().andThrow(exception);
+    mockWriter.write(anyObject());
+    expectLastCall().times(2);
+    expectLastCall().andThrow(exception);
```

Comment:
Do you mean testing that if there are errant records in the middle of a batch, good records before and after will be still sent to the database? Because in that case, yes this is tested in an integration test.

Reply:
I see where that is tested in an IT. But the IT is testing more than just this specific case. If it fails, it will be less obvious why it failed. In general, unittest coverage should be broader than IT coverage and there may well be overlapping coverage, which is fine.

Reply:
I've added a unit test to test that even if the bad record is in the middle of the batch, the good records are not sent to the error reporter. Let me know what you think